arch/arm/src/cxd56xx/cxd56_sfc.c

```diff
@@ -200,6 +200,7 @@ static int cxd56_ioctl(FAR struct mtd_dev_s *dev, int cmd, unsigned long arg)
               geo->blocksize    = PAGE_SIZE;
               geo->erasesize    = SECTOR_SIZE;
               geo->neraseblocks = priv->density >> SECTOR_SHIFT;
+              geo->firstblock   = 0;
```

Comment:
is it start better than first?

Reply:
Is firstblock even meaningful with some partition types.  I think mixing partitions and volume geometry is not a good idea.

Rather, it would make more sense to provide more complete partition information as you get from a partition tool like fdisk.  Again, we need to think of a roadmap.  Where do you want to go with this?  I don't think this gets your there.  It mixes functionality and cannot be built upon for more general partition support.

Reply:
add a new IOCTL MTDIOC_PARTINFO?

Reply:
> Rather, it would make more sense to provide more complete partition information as you get from a partition tool like fdisk.

Initially the idea of a new `MTDIOC_FIRSTBLOCK` ioctl command was to isolate the position from the geometry information, as from our (me and @Ouss4) interpretation it didn't make much sense to be together. But I had no stronger arguments to defend it.

> add a new IOCTL MTDIOC_PARTINFO?

It could be an option, but then much of the information to be retrieved via `MTDIOC_PARTINFO` would overlap with `MTDIOC_GEOMETRY`, being currently the only difference the `firstblock` info.

Reply:
It would make more sense to me, for example, if the geometry included a partition number perhaps.  Then if there were APIs (or procfs/partition) to view the partitions, then all other information about the partition could be obtained from the more extensive partition table.

Parititions all work differently.  The Microsoft-ish partition tables used with SD cards and FAT aver very different from MTD partitions.  Xaiomi added partition support of some kind too.  It seems like a big mistake to put partition information into volume information.  These are completely unrelated.

Reply:
> Initially the idea of a new `MTDIOC_FIRSTBLOCK` ioctl command was to isolate the position from the geometry information, as from our (me and @Ouss4) interpretation it didn't make much sense to be together. But I had no stronger arguments to defend it.

I think you are correct.  This is why a roadmap is critical.  You cannot evaluate technical decisions today if you don't know where you want to be tomorrow.

"Tomorrow" we will want better partitiion support.  We must plan for that today.  Mixing volume and partition information is a horrible idea.  We should not do that.

Reply:
> It would make more sense to me, for example, if the geometry included a partition number perhaps. Then if there were APIs (or procfs/partition) to view the partitions, then all other information about the partition could be obtained from the more extensive partition table.

Sorry, I couldn't follow your idea. By including "a partition number" you mean like an index? Or the number of partitions for the MTD?
The latter seems more reasonable, so that it would be possible to iterate the partition list, similar to what Linux has for MTD erase regions.
@xiaoxiang781216's suggestions for an `MTDIOC_PARTINFO` could also fit for when the caller already contains a direct reference to the partition (e.g. via a chardev path).

So, for an initial implementation, this could be the structure for a partition info:
```c
struct mtd_partition_info_s
{
  size_t          nblocks;       /* Number of blocks in the partition */
  off_t           startblock;    /* Offset to the first block of the managed
                                  * sub-region */
#ifdef CONFIG_MTD_PARTITION_NAMES
  FAR const char *name;          /* Name of the partition */
#endif
};
```
For consistency, `part_ioctl` would then always forward the handling of `MTDIOC_GEOMETRY` to the parent MTD.
The application would now have the responsibility for calculating the partition size by invoking both `MTDIOC_GEOMETRY` (for `blocksize`) and `MTDIOC_PARTINFO` (for `nblocks`).
Is this approach reasonable?

Reply:
Another, unrelated partition design problem is the way that the Microsoft DOS partitions are managed (e.g. on SD cards).  There is no standalone partition logic; the partition logic is built into the FAT file system.  If the first block is not the start of the FAT file system, then it scans the partition table and mounts the first FAT file system if finds.

That raises some problems.  It means, for example, that we cannot handle two FAT file systems in different partitions and that it would be be awkward to use the other partitions for anything else.

Reply:
This think I replied to the wrong address.  I will copy my response here if this is a duplicate and my email responses does show up, I will delete this:

> > It would make more sense to me, for example, if the geometry included a partition number perhaps. Then if there were APIs (or procfs/partition) to view the partitions, then all other information about the partition could be obtained from the more extensive partition table.
>
> Sorry, I couldn't follow your idea. By including "a partition number" you mean like an index? Or the number of partitions for the MTD?

Yes, I was thinking an index.  I am not sure if that is really that meaningful.

>
> So, for an initial implementation, this could be the structure for a partition info:
> struct mtd_partition_info_s
> {
>   size_t          nblocks;       /* Number of blocks in the partition */
>   off_t           startblock;    /* Offset to the first block of the managed
>                                   * sub-region */
> #ifdef CONFIG_MTD_PARTITION_NAMES
>   FAR const char *name;          /* Name of the partition */
> #endif
> };

Think of what you see if you list partitions with fdisk or other partition tool.  You would want, for example, the filesystem type if there is a file system in the partition.  This comes from the magic number at the beginning of the partition (as in include/sys/statfs.h).

With fdisk -l you would see the device, an indication of the volume is bootable, start sector, end sector, size in setions, size in human-readable bytes, ID (https://www.win.tue.nl/~aeb/partitions/partition_types-1.html), and human readable type.

Bootable does not apply to NuttX, at least at present.

I believe that reading the partition table from a block device is a user application.  Nothing special is required from the OS since the partition table can be directly read from the block device by a user-space application.

MTD makes this more complex, however, since MTD devices to do not have partition tables; mtd_partition.c is a software layer that imposes the partition layout on the device with nothing  stored on the media.  The good part of that design is that it is very light weight; the bad part is that only the firmware knows the partition layout of the FLASH memory.  There is no way to move an MTD part from on platform to another while retaining the partition information.  That should probably be made more standard someday.

Reply:
> MTD makes this more complex, however, since MTD devices to do not have partition tables; mtd_partition.c is a software layer that imposes the partition layout on the device with nothing stored on the media. The good part of that design is that it is very light weight; the bad part is that only the firmware knows the partition layout of the FLASH memory. There is no way to move an MTD part from on platform to another while retaining the partition information. That should probably be made more standard someday.

I should also mention..

For the case of MTD partitions, there is procfs support enabled by not excluding it with CONFIG_PROCFS_EXCLUDE_PARTITIONS.  Ken Pettit contributed this ages ago but I know nearly nothing about this, but believe that procfs is the best way to read the partition table.  It really shouldn't matter what kind of partition the device holds, as long as the procfs partition table format is the same.

This is why I say that if you know a partition table index, the you have complete knowledge of the partition... at least for the case of MTD.

Reply:
> Think of what you see if you list partitions with fdisk or other partition tool. You would want, for example, the filesystem type if there is a file system in the partition. This comes from the magic number at the beginning of the partition (as in include/sys/statfs.h).
> 
> With fdisk -l you would see the device, an indication of the volume is bootable, start sector, end sector, size in setions, size in human-readable bytes, ID (https://www.win.tue.nl/~aeb/partitions/partition_types-1.html), and human readable type.

> 
> MTD makes this more complex, however, since MTD devices to do not have partition tables; mtd_partition.c is a software layer that imposes the partition layout on the device with nothing stored on the media. The good part of that design is that it is very light weight; the bad part is that only the firmware knows the partition layout of the FLASH memory. There is no way to move an MTD part from on platform to another while retaining the partition information. That should probably be made more standard someday.

Ok, so for an MTD partition the proposed struct seems fit, maybe just an `endblock` field and change the `nblocks` to a `length` in bytes.

> I believe that reading the partition table from a block device is a user application. Nothing special is required from the OS since the partition table can be directly read from the block device by a user-space application.

Agreed. The information from a single partition (and the partition table, as well) can be retrieved both via IOCTL calls or via `procfs`. For the use case I am dealing with, an `ioctl` call seems more suitable. I believe the `procfs` approach fits best from NSH usage.

So your idea is that we should have a generic `struct partition_info`, not specific to MTD. Basically that would be the previous `mtd_partition_info` with added info, something like:
```c
struct mtd_partition_info_s
{
  uint32_t       magic;        /* File-system magic (0 for RAW) */
  size_t          length;        /* Partition size in bytes */
  size_t          numsectors;        /* Number of sectors in the partition  */
  off_t           startsector;     /* Offset to the first block/section of the managed
                                   * sub-region */
  off_t           endsector;     /* Offset to the last block of the managed
                                   * sub-region */
#ifdef CONFIG_MTD_PARTITION_NAMES
  FAR const char *name;           /* Name of the partition */
#endif
};
```

Reply:
I've updated the PR with my interpretation from recent suggestions.